File: scripts/calibration/scoreboard.py
#!/usr/bin/env python3
"""Compare packed estimates against production nvidia-smi totals.

Runs :mod:`dump_estimates` and joins its per-device placement against the
``prod-*`` rows in ``data/measurements.ndjson``, printing the drift for each
model. This is the top-level pass/fail signal for the ±5% estimation campaign.

Usage::

    python3 scripts/calibration/scoreboard.py
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
MEASUREMENTS = HERE / "data" / "measurements.ndjson"

# Maps a `models.toml` entry name to the `factors.label` of its production cell.
PROD_LABELS = {
    "qwen3.6-35b-a3b": "prod-qwen36-35b-a3b",
    "qwen3.6-27b": "prod-qwen36-27b",
    "gemma-4-31b-it-qat": "prod-gemma4-31b-qat",
    "deepseek-v4-flash": "prod-dsv4f",
    "glm-5.2": "prod-glm52",
    "laguna-s-2.1-iq4-nl": "prod-laguna",
    "talkie-1930-13b-it": "prod-talkie",
}

TOLERANCE_PCT = 5.0


def production_totals() -> dict[str, tuple[int, str]]:
    """The most recent production reading per label, with when it was taken.

    Most recent, not first: a runtime upgrade changes what the same
    configuration costs, and the old row then describes a different program.
    ik_llama's DSA compute buffer shrank by a third between two builds — GLM-5.2
    measured 11524 MiB of it on one and 7794 on the next, at byte-identical
    settings — so a stale row would have the estimator chasing a number the
    installed runtime no longer produces. A re-measurement under
    `prod-<name>-<suffix>` supersedes the original by date.
    """
    totals: dict[str, tuple[int, str]] = {}
    with MEASUREMENTS.open() as handle:
        for line in handle:
            row = json.loads(line)
            label = row["factors"].get("label", "")
            if not label.startswith("prod-") or row["status"] != "ok":
                continue
            when = row["provenance"]["measured_at_utc"]
            # A suffixed re-measurement counts for the label it re-measures.
            for base in {label, label.rsplit("-", 1)[0]} & set(PROD_LABELS.values()):
                if base not in totals or when > totals[base][1]:
                    totals[base] = (row["rss"]["gpu_used_mib"], when)
    return totals


def main() -> int:
    raw = subprocess.run(
        [sys.executable, str(HERE / "dump_estimates.py"), "--json"],
        capture_output=True,
        text=True,
        check=True,
    ).stdout
    estimates = {row["name"]: row for row in json.loads(raw)}
    prod = production_totals()

    print(f"{'Model':<24} {'Est GPU':>9} {'Prod GPU':>9} {'Drift':>8}  measured")
    worst = 0.0
    for name, label in PROD_LABELS.items():
        est_row = estimates.get(name)
        entry = prod.get(label)
        measured, when = entry if entry else (None, "")
        if est_row is None or measured is None:
            print(f"{name:<24} {'-':>9} {'-':>9} {'missing':>8}")
            continue
        placement = est_row["placement"]
        est = sum(v for k, v in placement.items() if k.startswith("gpu:"))
        drift = 100.0 * (est - measured) / measured
        flag = "" if abs(drift) <= TOLERANCE_PCT else "  <-- FAIL"
        worst = max(worst, abs(drift))
        print(f"{name:<24} {est:>9} {measured:>9} {drift:>+7.1f}%  {when[:10]}{flag}")
    print(f"\nworst drift: {worst:.1f}% (tolerance {TOLERANCE_PCT}%)")
    return 0 if worst <= TOLERANCE_PCT else 1


if __name__ == "__main__":
    sys.exit(main())

File: ananke-calibrate/src/bin/scoreboard.rs
//! Compare packed estimates against production `nvidia-smi` totals.
//!
//! The top-level pass/fail signal for the estimation campaign: seven models the
//! operator actually serves, each estimated the way the daemon would and compared
//! against what the driver reported for the running process.
//!
//! The estimator and the packer run in-process, once per model.

use std::{collections::BTreeMap, path::Path, process::ExitCode};

use ananke_calibrate::{
    models,
    record::read_ndjson,
    validate::{NEUTRAL, snapshot_for},
};
use ananke_fs::LocalFs;
use ananke_placement::pack_demand;

const MEASUREMENTS: &str = "scripts/calibration/data/measurements.ndjson";
const MODELS_TOML: &str = "scripts/calibration/models.toml";
const TOLERANCE_PCT: f64 = 5.0;

/// A `models.toml` entry name against the `factors.label` of its production cell.
const PROD_LABELS: &[(&str, &str)] = &[
    ("qwen3.6-35b-a3b", "prod-qwen36-35b-a3b"),
    ("qwen3.6-27b", "prod-qwen36-27b"),
    ("gemma-4-31b-it-qat", "prod-gemma4-31b-qat"),
    ("deepseek-v4-flash", "prod-dsv4f"),
    ("glm-5.2", "prod-glm52"),
    ("laguna-s-2.1-iq4-nl", "prod-laguna"),
    ("talkie-1930-13b-it", "prod-talkie"),
];

fn main() -> ExitCode {
    let text = match std::fs::read_to_string(MEASUREMENTS) {
        Ok(t) => t,
        Err(e) => {
            eprintln!("reading {MEASUREMENTS}: {e}");
            return ExitCode::from(2);
        }
    };
    let records = match read_ndjson(&text) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("parsing {MEASUREMENTS}: {e}");
            return ExitCode::from(2);
        }
    };
    let configs = match models::load(Path::new(MODELS_TOML)) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("{e}");
            return ExitCode::from(2);
        }
    };

    // The most recent reading per label, not the first. A runtime upgrade changes
    // what the same configuration costs, and the old row then describes a
    // different program: ik's DSA compute buffer shrank by a third between two
    // builds — GLM-5.2 measured 11524 MiB of it on one and 7794 on the next, at
    // byte-identical settings — so a stale row would have the estimator chasing a
    // number the installed runtime no longer produces. A re-measurement under
    // `prod-<name>-<suffix>` supersedes the original by date.
    let mut production: BTreeMap<&str, (u64, String)> = BTreeMap::new();
    for record in &records {
        let label = record.factors.label.as_str();
        if !label.starts_with("prod-") || record.status != "ok" {
            continue;
        }
        let Some(used) = record.gpu_used_mib() else {
            continue;
        };
        let when = record.provenance.measured_at_utc.clone();
        // A suffixed re-measurement counts for the label it re-measures.
        let base = label
            .rsplit_once('-')
            .map(|(head, _)| head)
            .unwrap_or(label);
        for candidate in [label, base] {
            if let Some((_, target)) = PROD_LABELS.iter().find(|(_, l)| *l == candidate) {
                let entry = production.entry(*target).or_insert((used, when.clone()));
                if when > entry.1 {
                    *entry = (used, when.clone());
                }
            }
        }
    }

    let fs = LocalFs;
    println!(
        "{:<24} {:>9} {:>9} {:>8}  measured",
        "Model", "Est GPU", "Prod GPU", "Drift"
    );
    let mut worst = 0.0_f64;
    for (name, label) in PROD_LABELS {
        let Some(config) = configs.iter().find(|c| c.name == *name) else {
            println!("{name:<24} {:>9} {:>9} {:>8}", "-", "-", "missing");
            continue;
        };
        let Some((measured, when)) = production.get(*label) else {
            println!("{name:<24} {:>9} {:>9} {:>8}", "-", "-", "missing");
            continue;
        };
        let model = config.model_path();
        let mmproj = config.mmproj_path();
        let draft = config.draft_path();
        let inputs = config.estimator_inputs(&model, mmproj.as_deref(), draft.as_deref());
        let estimate = match ananke_estimate::estimate_from_path(&fs, &inputs) {
            Ok(e) => e,
            Err(e) => {
                println!("{name:<24} estimator refused: {e}");
                continue;
            }
        };
        let placement = config.placement_inputs();
        let snap = snapshot_for(&placement.gpu_allow, &[]);
        let packed = match pack_demand(&estimate, &placement, &snap, NEUTRAL) {
            Ok(p) => p,
            Err(e) => {
                println!("{name:<24} packer refused: {e:?}");
                continue;
            }
        };
        // The reservation summed over the placement's GPU slots — the figure the
        // operator's cards actually hold.
        let est = ananke_calibrate::validate::gpu_reserved_mib(&packed);
        let drift = 100.0 * (est as f64 - *measured as f64) / *measured as f64;
        let flag = if drift.abs() <= TOLERANCE_PCT {
            ""
        } else {
            "  <-- FAIL"
        };
        worst = worst.max(drift.abs());
        println!(
            "{name:<24} {est:>9} {measured:>9} {drift:>+7.1}%  {}{flag}",
            &when[..when.len().min(10)]
        );
    }
    println!("\nworst drift: {worst:.1}% (tolerance {TOLERANCE_PCT:.0}%)");
    if worst <= TOLERANCE_PCT {
        ExitCode::SUCCESS
    } else {
        ExitCode::from(1)
    }
}
